File: typescript/examples/warehouse.ts
import { L2S1, type DecisionRequest } from '@l2s1/node';

const request: DecisionRequest = {
  state: { storage_requirement: 'chilled' },
  decisions: [{
    id: 'storage_zone', instruction: 'Select the storage zone matching storage_requirement.',
    kind: { type: 'choice', options: [
      { id: 'ambient', criterion: 'Ambient storage is required.' },
      { id: 'chilled', criterion: 'Chilled storage is required.' },
      { id: 'frozen', criterion: 'Frozen storage is required.' },
    ] },
  }],
};
const model = process.argv[2];
if (!model) throw new Error('Usage: node examples/warehouse.ts /path/to/model.gguf');
const engine = await L2S1.load({ model, ...(process.env.L2S1_BINARY ? { binaryPath: process.env.L2S1_BINARY } : {}) });
try { console.log(JSON.stringify(await engine.decide(request), null, 2)); }
finally { await engine.close(); }

File: examples/warehouse.rs
//! Construct and validate all three decision kinds directly in Rust.
use l2s1::{Decision, DecisionKind, DecisionRequest, Level, OptionSpec};
use serde_json::{Map, Value};

fn main() -> l2s1::Result<()> {
    let request = DecisionRequest {
        state: Value::Object(Map::from_iter([
            ("shipment_id".into(), Value::from("BOX-103")),
            ("storage_requirement".into(), Value::from("chilled")),
            ("hours_until_dispatch".into(), Value::from(4)),
        ])),
        decisions: vec![
            Decision {
                id: "storage_zone".into(),
                instruction: "Select the storage zone that matches the shipment's storage_requirement.".into(),
                kind: DecisionKind::Choice {
                    options: vec![
                        OptionSpec {
                            id: "ambient".into(),
                            criterion: "The shipment requires ambient storage.".into(),
                        },
                        OptionSpec {
                            id: "chilled".into(),
                            criterion: "The shipment requires chilled storage.".into(),
                        },
                        OptionSpec {
                            id: "frozen".into(),
                            criterion: "The shipment requires frozen storage.".into(),
                        },
                    ],
                },
            },
            Decision {
                id: "cold_chain_required".into(),
                instruction: "Does this shipment need temperature-controlled storage? Chilled and frozen shipments do; ambient shipments do not.".into(),
                kind: DecisionKind::Binary {
                    false_label: "No temperature control is required.".into(),
                    true_label: "Temperature control is required.".into(),
                },
            },
            Decision {
                id: "dispatch_priority".into(),
                instruction: "Choose the priority using hours_until_dispatch and the exact thresholds in the levels.".into(),
                kind: DecisionKind::Ordinal {
                    levels: vec![
                        Level {
                            id: "low".into(),
                            criterion: "More than 24 hours remain until dispatch.".into(),
                            value: 0.0,
                        },
                        Level {
                            id: "medium".into(),
                            criterion: "More than 6 hours and at most 24 hours remain until dispatch.".into(),
                            value: 1.0,
                        },
                        Level {
                            id: "high".into(),
                            criterion: "At most 6 hours remain until dispatch.".into(),
                            value: 2.0,
                        },
                    ],
                },
            },
        ],
    };
    request.validate()?;
    println!("{request:#?}");
    Ok(())
}
